packages/block-editor/src/components/block-tools/selected-block-popover.js

```diff
@@ -102,6 +102,12 @@ function SelectedBlockPopover( {
 	// to it when re-mounting.
 	const initialToolbarItemIndexRef = useRef();
 
+	useEffect( () => {
+		// Resets the index whenever the active block changes so this is not
+		// persisted. See https://github.com/WordPress/gutenberg/pull/25760#issuecomment-717906169
+		initialToolbarItemIndexRef.current = undefined;
+	}, [ clientId ] );
```

Comment:
Changing the `key` prop on a React element basically tells it that it's a different component now, triggering the component to re-mount. So the `BlockContextualToolbar` gets re-mounted when you change blocks.

And refs, when used like this, can prevent a component from re-rendering. So it prevents `SelectedBlockPopover` from re-rendering when the selected toolbar index changes.

So I guess it makes sense to me that they're both needed when I think through it. You can't move the state inside `BlockContextualToolbar` unless the `shouldShowContextualToolbar` state moves inside it too because the `BlockContextualToolbar` also gets remounted when the toolbar is hidden and then shown.

Maybe it would make sense to move that state inside to clear up some of the weirdness, but this PR is probably good enough for now to get the fix in.

Reply:
@ajlende and I worked on this refactor and decided to leave it as is, as the refactor was getting too involved.